statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,436 @@
+debiasing procedure.
+
+    X_beta
+
+    A weighted design matrix used to perform the node-wise
+    regression procedure.
+
+    gamma_hat
+
+    gamma_hat is produced as part of the node-wise regression
+    procedure used to produce the approximate inverse covariance
+    matrix.  One is produced for each variable using the
+    LASSO.
+
+    tau_hat
+
+    tau_hat is produced using the gamma_hat values for each
+    p to produce weights to reweight the gamma_hat values which
+    are ultimately used to form theta_hat.
+
+    theta_hat
+
+    This is the estimate of the approximate inverse covariance
+    matrix.  This is used to debiase the coefficient average
+    along with the average gradient.  For the OLS case,
+    theta_hat is an approximation for
+
+        (1 / n X^T X)^{-1}
+
+    formed by node-wise regression.
+"""
+
+def _gen_grad(mod, params, alpha, L1_wt, score_kwds):
+    """generates the log-likelihood gradient for the debiasing
+
+    Parameters
+    ----------
+    mod : statsmodels model class
+        The model for the current machine.
+    params : array-like
+        The estimated coefficients for the current machine.
+    alpha : scalar or array-like
+        The penalty weight.  If a scalar, the same penalty weight
+        applies to all variables in the model.  If a vector, it
+        must have the same length as `params`, and contains a
+        penalty weight for each coefficient.
+    L1_wt : scalar
+        The fraction of the penalty given to the L1 penalty term.
+        Must be between 0 and 1 (inclusive).  If 0, the fit is
+        a ridge fit, if 1 it is a lasso fit.
+    score_kwds : dict-like or None
+        Keyword arguments for the score function.
+
+    Returns
+    -------
+    An array-like object of the same dimension as params
+
+    Notes
+    -----
+    In general:
+
+    gradient l_k(params)
+
+    where k corresponds to the index of the machine
+
+    For the simple linear case:
+
+    X^T(y - X^T params)
+    """
+
+    grad = -mod.score(np.r_[params], **score_kwds)
+    # this second part comes from the elastic net penalty
+    grad += alpha * (1 - L1_wt)
+    return grad
+
+
+def _gen_wdesign_mat(mod, params, hess_kwds):
+    """generates the weighted design matrix necessary to generate
+    the approximate inverse covariance matrix
+
+    Parameters
+    ----------
+    mod : statsmodels model class
+        The model for the current machine.
+    params : array-like
+        The estimated coefficients for the current machine.
+    hess_kwds : dict-like or None
+        Keyword arguments for the hessian function.
+
+    Returns
+    -------
+    An array-like object, updated design matrix, same dimension
+    as mod.exog
+    """
+
+    # TODO need to handle duration and other linear model classes
+    rhess = np.sqrt(mod.hessian_obs(np.array(params), **hess_kwds))
+    return rhess[:, None] * mod.exog
+
+
+def _gen_gamma_hat(X_beta, pi, alpha):
```

Comment:
Yes, wexog is good, we have used it elsewhere.

On Mon, Jun 27, 2016 at 1:25 PM, Leland Bybee <notifications@github.com>
wrote:

> In statsmodels/base/distributed_estimation.py
> <https://github.com/statsmodels/statsmodels/pull/3055#discussion_r68618829>
> :
>
> > +        The estimated coefficients for the current machine.
> > +    hess_kwds : dict-like or None
> > +        Keyword arguments for the hessian function.
> > +
> > +    Returns
> > +    -------
> > +    An array-like object, updated design matrix, same dimension
> > +    as mod.exog
> > +    """
> > +
> > +    # TODO need to handle duration and other linear model classes
> > +    rhess = np.sqrt(mod.hessian_obs(np.array(params), **hess_kwds))
> > +    return rhess[:, None] * mod.exog
> > +
> > +
> > +def _gen_gamma_hat(X_beta, pi, alpha):
>
> I'm not sure if exog makes the most sense as a replacement for X_beta
> since it isn't necessarily the same as the model exog. Something like
> wexog or weighted_exog might make more sense as a replacement. I do agree
> with getting away from the greek letters though, and I like nodewise_row,
> I'll be making these changes with the next round of commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/statsmodels/statsmodels/pull/3055/files/8ccb5a677757d76222facf45f2550709fc9388cb#r68618829>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACiww1Xd4N8Fr6vz97f4HsSYHygcm6tfks5qQAeTgaJpZM4I6Q4M>
> .
>